services/wfm/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/CommandBuilderImpl.java

```diff
@@ -136,21 +138,44 @@ public CommandBuilderImpl(PersistenceManager persistenceManager, FlowResourcesCo
                 .collect(Collectors.toList());
     }
 
-    private RemoveFlow buildRemoveFlowWithoutMeterFromFlowEntry(SwitchId switchId, FlowEntry entry) {
+    @VisibleForTesting
+    RemoveFlow buildRemoveFlowWithoutMeterFromFlowEntry(SwitchId switchId, FlowEntry entry) {
         Optional<FlowMatchField> entryMatch = Optional.ofNullable(entry.getMatch());
+        Optional<FlowInstructions> instructions = Optional.ofNullable(entry.getInstructions());
+        Optional<FlowApplyActions> applyActions = instructions.map(FlowInstructions::getApplyActions);
 
         Integer inPort = entryMatch.map(FlowMatchField::getInPort).map(Integer::valueOf).orElse(null);
+
+        FlowEncapsulationType encapsulationType = FlowEncapsulationType.TRANSIT_VLAN;
+        Integer encapsulationId = null;
         Integer vlan = entryMatch.map(FlowMatchField::getVlanVid).map(Integer::valueOf).orElse(null);
+        if (vlan != null) {
+            encapsulationId = vlan;
+        } else {
+            Integer tunnelId = entryMatch.map(FlowMatchField::getTunnelId).map(Integer::valueOf).orElse(null);
```

Comment:
Can you leave optional and do not assign null value and then you can use optional check instead of comparing with a null